### PR TITLE
PrepareBeaconProposer API 

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

part of #10292

**What does this PR do? Why is it needed?**

This PR implements the inner workings of https://github.com/ethereum/beacon-APIs/pull/178/files Beacon API to set the prysm config. This cache is then used when calling get execution payload

### CODE REVIEW

File: beacon-chain/rpc/eth/validator/validator.go [hunk 1]
```diff
@@ -325,11 +326,22 @@ func (vs *Server) ProduceBlockV2(ctx context.Context, req *ethpbv1.ProduceBlockR
...
-	_ context.Context, _ *ethpbv1.PrepareBeaconProposerRequest,
+	ctx context.Context, request *ethpbv1.PrepareBeaconProposerRequest,
 ) (*emptypb.Empty, error) {
-	return &emptypb.Empty{}, status.Error(codes.Unimplemented, "Unimplemented")
+	_, span := trace.StartSpan(ctx, "validator.PrepareBeaconProposer")
+	defer span.End()
+	var FeeRecipients []common.Address
+	var ValidatorIndices []types.ValidatorIndex
+	for _, recipientContainer := range request.Recipients {
+		FeeRecipients = append(FeeRecipients, common.BytesToAddress(recipientContainer.FeeRecipient))
```
Comment (terencechain) [inline beacon-chain/rpc/eth/validator/validator.go:338]: I think it's worth validating `common.BytesToAddress(recipientContainer.FeeRecipient)`

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 2]
```diff
@@ -116,11 +118,22 @@ func (vs *Server) ProposeBlock(ctx context.Context, rBlk *ethpb.SignedBeaconBloc
...
-	_ context.Context, _ *ethpb.PrepareBeaconProposerRequest,
+	ctx context.Context, request *ethpb.PrepareBeaconProposerRequest,
 ) (*emptypb.Empty, error) {
-	return &emptypb.Empty{}, status.Error(codes.Unimplemented, "Unimplemented")
+	_, span := trace.StartSpan(ctx, "validator.PrepareBeaconProposer")
+	defer span.End()
+	var FeeRecipients []common.Address
+	var ValidatorIndices []types.ValidatorIndex
+	for _, recipientContainer := range request.Recipients {
+		FeeRecipients = append(FeeRecipients, common.BytesToAddress(recipientContainer.FeeRecipient))
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:130]: Same here. Worth validating `common.BytesToAddress(recipientContainer.FeeRecipient)`

File: beacon-chain/rpc/eth/validator/validator.go [hunk 3]
```diff
@@ -325,11 +326,25 @@ func (vs *Server) ProduceBlockV2(ctx context.Context, req *ethpbv1.ProduceBlockR
...
-	_ context.Context, _ *ethpbv1.PrepareBeaconProposerRequest,
+	ctx context.Context, request *ethpbv1.PrepareBeaconProposerRequest,
 ) (*emptypb.Empty, error) {
-	return &emptypb.Empty{}, status.Error(codes.Unimplemented, "Unimplemented")
+	_, span := trace.StartSpan(ctx, "validator.PrepareBeaconProposer")
+	defer span.End()
+	var FeeRecipients []common.Address
+	var ValidatorIndices []types.ValidatorIndex
+	for _, recipientContainer := range request.Recipients {
+		if len(recipientContainer.FeeRecipient) != fieldparams.FeeRecipientLength {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/rpc/eth/validator/validator.go:338]: I can do it that way too will update then.

File: beacon-chain/rpc/eth/validator/validator_test.go [hunk 4]
```diff
@@ -2011,3 +2013,62 @@ func TestSubmitContributionAndProofs(t *testing.T) {
...
+							ValidatorIndex: 1,
+						},
+					},
+				},
+			},
+			wantErr: status.Errorf(codes.InvalidArgument, "Invalid fee recipient length"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/rpc/eth/validator/validator_test.go:2054]: This test should also read from the DB to ensure the expected data was saved

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_test.go [hunk 5]
```diff
@@ -2431,6 +2432,59 @@ func TestProposer_GetSyncAggregate_OK(t *testing.T) {
...
+			},
+			wantErr: status.Errorf(codes.InvalidArgument, "Invalid fee recipient length"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			db := dbutil.SetupDB(t)
+			ctx := context.Background()
+			proposerServer := &Server{BeaconDB: db}
+			_, err := proposerServer.PrepareBeaconProposer(ctx, tt.args.request)
```
Comment (rauljordan) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_test.go:2477]: +1 here

**Pull request merged**